### PR TITLE
Fixed use of memories in XLNet (caching for language generation + warning when loading improper memoryless model)

### PR DESCRIPTION
**Pull request by TevenLeScao**
The default XLNet model is loaded with 0 memory length, which is an issue both at training time (improper performance) and inference time (as there's no caching speed-up since it doesn't return former attentions). As discussed with @LysandreJik , this PR introduces a warning that in the future, the default XLNet model will have 1024 memory length, in accordance with [the original paper](https://arxiv.org/abs/1906.08237). It also fixes the re-use of cached memory, which was broken similarly to TransfoXL (#4752).

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
LGTM, thanks for taking care of it!

File: src/transformers/modeling_xlnet.py [hunk 1]
```diff
@@ -1000,7 +1000,15 @@ def prepare_inputs_for_generation(self, input_ids, past, **kwargs):
 
         effective_batch_size = input_ids.shape[0]
         dummy_token = torch.zeros((effective_batch_size, 1), dtype=torch.long, device=input_ids.device)
-        input_ids = torch.cat([input_ids, dummy_token], dim=1)
+
+        # at every pass, the attention values for the new token and the two last generated tokens
+        # are computed, the rest is reloaded from the `past` cache
+        offset = 2
```
Comment (patrickvonplaten) [inline src/transformers/modeling_xlnet.py:1006]: I looked a bit deeper into the code, especially this line:
https://github.com/huggingface/transformers/blob/b25f7802de2749a5f8c3430437eceabf9e6384b8/src/transformers/modeling_xlnet.py#L336


My conclusion is 
a) 
key, value are always recalculated for each memory position -> so that there is actually 0 logic in the code that speeds up attention. We also saw this in our timing experiments. To me it seems that key and value projections are recalculated for each past memory position even though this position was already calculated before and thanks to causal masking should not influence future calculations.
b) 
mems are completely different do `past` as it is in GPT2 and are *not* key, value projections, but hidden states (the outputs of attention layers) => so we might need to add a complete new `past_key_value` varable here to speed up inference. Inference is extremely slow in XLNet and we had many issues about it. Looking at the code, I think we can implement this and could gain significantly speed ups *with equal* mathematical output (not a 100% sure though - will have to dive deeper in the code...what do you think @TevenLeScao ?)
c)
The code as implemented by you and as discussed I think is logically the correct way to implement causal language modelling for XLNet, but setting `offset = 2` seems a bit random here...maybe we can even contact the author for more information how exactly causal lm should be implemented as setting offest just to 2 seems a bit random to me.

**Comment by patrickvonplaten** [comment]
Also the slow test of XLNet will have to be adapted when doing this change no?

**Comment by TevenLeScao** [comment]
So in order:

a) b) I would really like to to have proper GPT-2 style caching in XLNet. This would require changing the outputs of the XLNet `forward` to add `presents` outputs that contain the K/V pairs, like in GPT-2. That probably counts as BC-breaking, right @LysandreJik ?

c) `offset = 2` is definitely a bit random. I was mistaken at first, as I thought it would be proper autoregressive generation, but `offset=1` is what should be used in this case. After manually checking outputs I had the impression that `offset = 2` was slightly better (mostly it goes less into repetitive generation loops) at a negligible computation time cost; but I agree that `offset = 1` is more principled and I don't have a strong opinion on that choice.

d) The slow tests run with the default model, which still has `mem_length = 0` and no caching, so it doesn't make a difference yet.

**Comment by patrickvonplaten** [comment]
> So in order:
> 
> a) b) I would really like to to have proper GPT-2 style caching in XLNet. This would require changing the outputs of the XLNet `forward` to add `presents` outputs that contain the K/V pairs, like in GPT-2. That probably counts as BC-breaking, right @LysandreJik ?
> 
> c) `offset = 2` is definitely a bit random. I was mistaken at first, as I thought it would be proper autoregressive generation, but `offset=1` is what should be used in this case. After manually checking outputs I had the impression that `offset = 2` was slightly better (mostly it goes less into repetitive generation loops) at a negligible computation time cost; but I agree that `offset = 1` is more principled and I don't have a strong opinion on that choice.
> 
> d) The slow tests run with the default model, which still has `mem_length = 0` and no caching, so it doesn't make a difference yet.

a)b) I think would count as a feature enhancement. We would make the `past` variable optional so no backward breaking here IMO. But I agree it would definitely be better to this in a new PR.

**Review by patrickvonplaten — APPROVED** [review]
The line `offset = 2` deserves a comment IMO. Apart from that I'm good with the PR!

**Comment by TevenLeScao** [comment]
I've added a comment. I'm preparing another PR for proper caching and merging this one.
